proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,166 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+Exception Types
+
+```
+exntype ::= [t*] -> []
+```
+
+### Instructions
+
+```
+instr ::= ... | throw x | rethrow l
+        | try bt instr* (catch x instr*)* (catch_all instr*)? end
+        | try bt instr* delegate l
+        | try bt instr* unwind instr* end
+```
+
+### Modules
+
+Exceptions (definitions)
+
+```
+exn ::= export* exception exntype  | export* exception exntype import
+```
+
+Modules
+
+
+```
+mod ::= module ... exn*
+```
+
+## Validation (Typing)
+
+To verify that the `rethrow l` instruction refers to a surrounding catch block, we introduce a stack `caught` to validation contexts, which gets an exception index or the keywork `all` prepended whenever we enter instructions inside a `catch exnidx` or `catch_all` block, respectively. This addition is reflected in the execution rules, by the administrative instruction `caught` which models the stack of caught exceptions on the wasm stack.
+
+
+### Instructions
+
+
+
+```
+C_exn(x) = [t*] -> []
+--------------------------------
+C |- throw x : [t1* t*] -> [t2*]
+
+
+C_caught(l) =/= ε
+-------------------------------
+C |- rethrow l : [t1*] -> [t2*]
+
+
+bt = [t1*] -> [t2*]
+C, label [t2*] |- instr* : [t1*] -> [t2*]
+(C_exn(x_i) = [t'_i*] -> [])^n
+(C, label [t2*], caught x_i |- instr_i* : [t'_i*] -> [t2*])^n
+(C, label [t2*], caught all |- instr'* : [] -> [t2*])?
+------------------------------------------------------------------
+try bt instr* (catch x_i instr_i*)^n (catch_all instr'*)? end : bt
+
+
+bt = [t1*] -> [t2*]
+C, label [t2*] |- instr* : [t1*] -> [t2*]
+C_label(l) =/= ε
+-----------------------------------------
+try bt instr* delegate l : bt
```

Comment:
This seems we are using a regular labels for `delegate`. Can we handle this case then?
```wast
try $label0
catch
  try
    ...
  delegate $label0  ;; cannot be done, because $label0's catch is not below but above here
end
```

Is this in line with the @RossTate's [suggestion](https://github.com/WebAssembly/exception-handling/pull/143#discussion_r522996966)? His suggestion looks like we don't need to use labels at all. Am I interpreting it correctly, @RossTate?

Reply:
That's in line with my suggestion. The number just indicates how many syntactic enclosures to skip out of. A convention in the text format could be that you can label a syntactic enclosure and reference that label in the `delegate` rather than have to compute the number of syntactic enclosures between.

Reply:
@ioannad What do you think? If we are on the same page on this I'll update the explainer to match this.

Reply:
@aheejin, @RossTate, that sounds good to me!

@aheejin, the example you gave above is now correctly illegal (or nonsensical - not translatable to integer labels), because the additional label attributes are removed (that is, catch-labels are removed, which were wrong, and try blocks are back to using regular labels),

Reply:
Sorry what I meant was not catch label but try label. In my example the label `$label0` is attached to `try`. So it should be a valid try label, but the code itself should be invalid. I'm not sure how you can describe that code using try label, and that's the question I asked.

Did your last commit contain any changes related to this? I don't see any, so...

Reply:
@aheejin, apologies for the late response. Let me start by saying that I do agree that your example should be impossible. Thank you for clarifying, I see what the problem is, and I also see that we actually do need `caught` to have a label after all, in order to accommodate situations such as:

```
try bt
  throw $e
catch $e
  ....
  br 0 ;; ---
end   ;;    |
;; <------- break to here
```

Tricky situation, I'll have to think about how to express this formally.

Reply:
@ioannad Not sure if I understand. Why is `br 0` a problem here? `br` can use `try` labels just like `block` labels because `try` can also serve as a `block`. Did you try to write `delegate 0` instead in your last example?

Maybe we can just do away with labels for `delegate` definition as @RossTate suggested.